From f819688c6f1d9b37c111971bdcf3e6ea31d789fd Mon Sep 17 00:00:00 2001
From: Dan Villiom Podlaski Christiansen <dchristiansen@zimbra.com>
Date: Thu, 22 Jan 2015 14:27:33 +0100
Subject: [PATCH] Prevent handled tab events from propagating outside the
 editor.

Allow 'tab' to escape from non-modal elements. Otherwise, the
containing application might assume that it should handle it.

Currently, TinyMCE always propagates tab key events. Unfortunately,
containing applications have little way of knowing whether this is a
tab that they should intercept -- such as in the toolbar -- or a tab
they should most definitely ignore -- such as in menus and dialogs.

We handle this by only letting tab keys propagate when unhandled; in
order to prevent toolbars from becoming tab sinks, we also add a
notion of controls that wrap focus. The latter currently applies to
all float panels.

---
diff -aur -x node_modules /tmp/root/tinymce-4.9.11/src/ui/main/ts/FloatPanel.ts tinymce-4.9.11/src/ui/main/ts/FloatPanel.ts
--- /tmp/root/tinymce-4.9.11/src/ui/main/ts/FloatPanel.ts	2020-07-13 04:14:19.000000000 +0000
+++ tinymce-4.9.11/src/ui/main/ts/FloatPanel.ts	2022-08-10 06:04:58.811153106 +0000
@@ -192,6 +192,9 @@
 
 const FloatPanel = Panel.extend({
   Mixins: [Movable, Resizable],
+  Defaults: {
+    wrapFocus: true
+  },
 
   /**
    * Constructs a new control instance with the specified settings.
diff -aur -x node_modules /tmp/root/tinymce-4.9.11/src/ui/main/ts/KeyboardNavigation.ts tinymce-4.9.11/src/ui/main/ts/KeyboardNavigation.ts
--- /tmp/root/tinymce-4.9.11/src/ui/main/ts/KeyboardNavigation.ts	2022-08-10 06:09:30.596375969 +0000
+++ tinymce-4.9.11/src/ui/main/ts/KeyboardNavigation.ts	2022-08-10 06:11:45.710989591 +0000
@@ -226,6 +226,8 @@
    * @private
    * @param {Number} dir Direction to move in positive means forward, negative means backwards.
    * @param {Array} elements Optional array of elements to move within defaults to the current navigation roots elements.
+   *
+   * @return {Boolean} Whether focus moved.
    */
   function moveFocus(dir, elements?) {
     let idx = -1;
@@ -240,7 +242,14 @@
     }
 
     idx += dir;
+
+    if (!navigationRoot.settings.wrapFocus && (idx < 0 || idx >= elements.length)) {
+      return false;
+    }
+
     navigationRoot.lastAriaIndex = moveFocusToIndex(idx, elements);
+
+    return true;
   }
 
   /**
@@ -318,8 +327,10 @@
       if (elm) {
         elm.focus();
       }
+
+      return true;
     } else {
-      moveFocus(e.shiftKey ? -1 : 1);
+      return moveFocus(e.shiftKey ? -1 : 1);
     }
   }
 
@@ -384,6 +395,7 @@
         const cancelEv = cancel();
         if (cancelEv.isDefaultPrevented()) {
           e.preventDefault();
+          e.stopPropagation();
         }
         if (cancelEv.isPropagationStopped()) {
           e.stopPropagation();
@@ -397,8 +409,10 @@
         break;
 
       case 9: // DOM_VK_TAB
-        tab(e);
-        e.preventDefault();
+        if (tab(e) !== false) {
+          e.preventDefault();
+          e.stopPropagation();
+        }
         break;
     }
   });
diff -aur -x node_modules /tmp/root/tinymce-4.9.11/src/ui/main/ts/Window.ts tinymce-4.9.11/src/ui/main/ts/Window.ts
--- /tmp/root/tinymce-4.9.11/src/ui/main/ts/Window.ts	2022-08-10 06:09:30.600376047 +0000
+++ tinymce-4.9.11/src/ui/main/ts/Window.ts	2022-08-10 06:08:44.687482992 +0000
@@ -119,6 +119,7 @@
     layout: 'flex',
     containerCls: 'panel',
     role: 'dialog',
+    ariaRoot: true,
     callbacks: {
       submit () {
         this.fire('submit', { data: this.toJSON() });
